### PR TITLE
Tpetra: refactor atomic maximum absolute value of scalar (compatibility with Kokkos)

### PR DESCRIPTION
**Pull request by dalg24**
@csiefer2
@ndellingwood 

## Motivation
Fix kokkos/kokkos#5815

### CODE REVIEW

**Review by ndellingwood — APPROVED** [review]
Changes look reasonable, a member of @trilinos/tpetra should confirm for final approval

**Comment by ndellingwood** [comment]
@trilinos/tpetra

**Comment by ndellingwood** [comment]
Thanks for the fast PR @dalg24 !

**Review by crtrott — APPROVED** [review]

**Review by csiefer2 — APPROVED** [review]

**Review by ndellingwood — APPROVED** [review]

File: packages/tpetra/core/src/kokkos_refactor/Tpetra_KokkosRefactor_Details_MultiVectorDistObjectKernels.hpp [hunk 1]
```diff
@@ -802,7 +803,7 @@ outOfBounds (const IntegerType x, const IntegerType exclusiveUpperBound)
     template <typename SC>
     KOKKOS_INLINE_FUNCTION
     void operator() (nonatomic_tag, SC& dst, const SC& src) const {
-      if (Kokkos::abs(dst) < Kokkos::abs(src)) dst = src;
+      if (Kokkos::ArithTraits<Scalar>::abs(dst) < Kokkos::ArithTraits<Scalar>::abs(src)) dst = src;
```
Comment (dalg24) [inline packages/tpetra/core/src/kokkos_refactor/Tpetra_KokkosRefactor_Details_MultiVectorDistObjectKernels.hpp:806]:
```suggestion
      if (Kokkos::ArithTraits<SC>::abs(dst) < Kokkos::ArithTraits<SC>::abs(src)) dst = src;
```

Comment (ndellingwood) [reply packages/tpetra/core/src/kokkos_refactor/Tpetra_KokkosRefactor_Details_MultiVectorDistObjectKernels.hpp]: Ack, thanks for the catch, fixed in ed2d18e1ecdc31bd471ef550e193ed8498f4a9e0

**Review by ndellingwood — APPROVED** [review]

**Comment by fryeguy52** [comment]
@ndellingwood @csiefer2 @crtrott I added AT:RETEST because I suspect failures last night were due to system issues on ascic166

**Comment by ndellingwood** [comment]
Thanks @fryeguy52 !

**Comment by ndellingwood** [comment]
Looks like there were some legitimate failures in the `TpetraCore_ImportExport_UnitTests_MPI_4` test in various builds related to changes in the PR

```
    5. ImportExport_int_longlong_Kokkos_Compat_KokkosCudaWrapperNode_AbsMax_UnitTest ... 
    8. ImportExport_int_longlong_Kokkos_Compat_KokkosSerialWrapperNode_AbsMax_UnitTest ... 
```

```
5. ImportExport_int_longlong_Kokkos_Compat_KokkosCudaWrapperNode_AbsMax_UnitTest ... 
 importer->isLocallyFitted() = true == dmap->isLocallyFitted(*smap) = true : passed
 Comparing tuple<SC>(-1.0,3.0) == dstVec->get1dView() ... 
 Error, tuple<SC>(-1.0,3.0)[1] = 3 == dstVec->get1dView()[1] = -3: failed!
 globalSuccess_int = 4 == 0 : FAILED ==> /scratch/trilinos/workspace/Trilinos_PR_cuda-11.4.2-uvm-off/Trilinos/packages/tpetra/core/test/ImportExport/ImportExport_UnitTests.cpp:385
 [FAILED]  (0.011 sec) ImportExport_int_longlong_Kokkos_Compat_KokkosCudaWrapperNode_AbsMax_UnitTest
 Location: /scratch/trilinos/workspace/Trilinos_PR_cuda-11.4.2-uvm-off/Trilinos/packages/tpetra/core/test/ImportExport/ImportExport_UnitTests.cpp:351

8. ImportExport_int_longlong_Kokkos_Compat_KokkosSerialWrapperNode_AbsMax_UnitTest ... 
 importer->isLocallyFitted() = true == dmap->isLocallyFitted(*smap) = true : passed
 Comparing tuple<SC>(-1.0,3.0) == dstVec->get1dView() ... 
 Error, tuple<SC>(-1.0,3.0)[1] = 3 == dstVec->get1dView()[1] = -3: failed!
 globalSuccess_int = 4 == 0 : FAILED ==> /scratch/trilinos/workspace/Trilinos_PR_cuda-11.4.2-uvm-off/Trilinos/packages/tpetra/core/test/ImportExport/ImportExport_UnitTests.cpp:385
 [FAILED]  (0.000178 sec) ImportExport_int_longlong_Kokkos_Compat_KokkosSerialWrapperNode_AbsMax_UnitTest
 Location: /scratch/trilinos/workspace/Trilinos_PR_cuda-11.4.2-uvm-off/Trilinos/packages/tpetra/core/test/ImportExport/ImportExport_UnitTests.cpp:351
```

File: packages/tpetra/core/src/kokkos_refactor/Tpetra_KokkosRefactor_Details_MultiVectorDistObjectKernels.hpp [hunk 2]
```diff
@@ -786,28 +785,25 @@ outOfBounds (const IntegerType x, const IntegerType exclusiveUpperBound)
...
   };
 
-  // Kokkos::Impl::atomic_fetch_oper wants a class like this.
-  template<class Scalar1, class Scalar2>
-  struct AbsMaxOper {
-    KOKKOS_INLINE_FUNCTION
-    static Scalar1 apply(const Scalar1& val1, const Scalar2& val2) {
-      const auto val1_abs = Kokkos::ArithTraits<Scalar1>::abs(val1);
-      const auto val2_abs = Kokkos::ArithTraits<Scalar2>::abs(val2);
-      return val1_abs > val2_abs ? Scalar1(val1_abs) : Scalar1(val2_abs);
```
Comment (dalg24) [inline packages/tpetra/core/src/kokkos_refactor/Tpetra_KokkosRefactor_Details_MultiVectorDistObjectKernels.hpp:796]: @ndellingwood I misread what this does.  It actually returns the absolute value.  The code I proposed below preserves the sign...

**Review by ndellingwood — APPROVED** [review]

**Comment by csiefer2** [comment]
Consulted with @dalg24 and @crtrott will advise after figuring out why this code does what it does and if we really want that behavior.

**Comment by csiefer2** [comment]
As per discussions with @dalg24 and @crtrott , we'll merge this code as is, assuming it works.  However, we're not sure that the current behavior of Tpetra is actually what we want (there are mistakes in the documentation for instance).  I will write unit tests and speak with stakeholders to figure out what we actually want this code to do.

If we decide to change behavior, we will alert the Kokkos Team

**Comment by ndellingwood** [comment]
@csiefer2 is this PR safe to merge, or is additional non-CI testing needed as well?

**Pull request merged**